### ISSUE TITLE
Ctype.Unify failure with definition of 'sequence'

### ISSUE THREAD

**Issue by yallop** [ocamllabs/ocaml-modular-implicits#4]
The code:

``` ocaml
module type MONAD =
sig
  type +_ t
  val return : 'a -> 'a t
  val (>>=) : 'a t -> ('a -> 'b t) -> 'b t
end

let rec sequence : 'a. (implicit M : MONAD) -> 'a M.t list -> 'a list M.t =
  fun (implicit M : MONAD) -> let open M in function
  | [] -> return []
  | x :: xs -> x >>= fun y ->
           sequence xs >>= fun ys ->
           return (y :: ys)
```

The behaviour:

``` bash
$ ocamlc monads.ml
Fatal error: exception Ctype.Unify(_)
```

**Comment by let-def** [comment]
This fails be cause the `'a` of the constraint is still a univar in the recursive call to `sequence`.
I don't know enough about univars and poly to tell why this is happening, my understanding was that it should have been instantiated to a var.
@lpw25 any idea?

**Comment by lpw25** [comment]
I've partially fixed this on [my branch](https://github.com/lpw25/ocaml/tree/modular-implicits). The univars were appearing because we weren't always generalizing implicits before doing regular generalization. 

I also found a type marking bug which was causing similar definitions of sequence to fail. There is still another bug preventing related to recursive calls that I hope to track down tomorrow.

**Comment by lpw25** [comment]
This is fixed by forward type propagation (#19).
